Perform all lookups using a configurable LookupEnv function

thallgren opened the pull request:
Adds the new type `type LookupEnv func(key string) (string, bool)` and a new `StructParser.ParseUsingLookup` method that takes a `LookupEnv`. argument. This enables tests to use this package without concern of concurrency issues with global environment variables.

thallgren committed:
Perform all lookups using a configurable LookupEnv function

Adds the new type `type LookupEnv func(key string) (string, bool)` and
a new `StructParser.ParseUsingLookup` method that takes a `LookupEnv`.
argument. This enables tests to use this package without concern of
concurrency issues with global environment variables.

LukeShu commented on envconfig.go:
```diff
@@ -272,6 +276,12 @@ func generateFieldHandler(i int, tag envTag, typeHandler FieldTypeHandler) func(
+
+// ParseUsingLookup populates structPtr from values returned by the given LookupEnv function, returning warnings and
+// fatal errors. It panics if structPtr is of the wrong type for this parser.
+func (p StructParser) ParseUsingLookup(structPtr interface{}, lookup LookupEnv) (warn, fatal []error) {
```
TBH, I'd just have this replace `ParseFromEnv`. Especially since I'm worried that #1 is a breaking change for apro.git; emphasize that there are breaking changes going in.

Nit on naming: `s/Using/From/` it's parsing the results from `Lookup`; it's not using `Lookup` to parse something.

LukeShu commented on envconfig.go:
```diff
@@ -77,6 +77,10 @@ func stringPointer(str string) *string {
 
+// LookupEnv is a function that performs lookup of an environment variable. It's typically
+// set to os.LookupEnv.
+type LookupEnv func(key string) (string, bool)
```
This type name is weird to me. Perhaps `LookupFunc`?

thallgren committed:
Changes in response to code review

- Rename the `LookupEnv` type to `LookupFunc`
- Let `ParseUsingLookup` replace the current `ParseFromEnv`

LukeShu reviewed (approved):
Woohoo!
